"Precondition failed: UID in precondition" flakes

**Which jobs are failing**:

pull-kubernetes-integration

**Which test(s) are failing**:

https://k8s-testgrid.appspot.com/google-unit#integration-master&include-filter-by-regex=apiextensions-apiserver%2Ftest%2Fintegration&width=5&sort-by-flakiness=

`./vendor/k8s.io/apiextensions-apiserver/test/integration/subresources_test.go -run TestValidateOnlyStatus`

`./vendor/k8s.io/apiextensions-apiserver/test/integration -run TestGeneration`

`./vendor/k8s.io/apiextensions-apiserver/test/integration/subresources_test.go -run TestStatusSubresource`

Not showing up in triage, oddly, but seen in individual failures:

https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/80485/pull-kubernetes-integration/1167014135116861443/

https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-integration/1167057042045669376

https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-integration/1167053904014217218

https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-integration/1167014135116861443

https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-integration/1166986078263971840

https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-integration-master/1165810482242654212

https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-integration-master/1166792805272522752

not sure if this specific to custom resources, but that's where I'm seeing the flakes

/sig api-machinery
/priority important-soon
/area custom-resources

talked with @liggitt offline, we get it reproduced by adding some delay to the watch cache:
```
diff --git a/staging/src/k8s.io/apiextensions-apiserver/test/integration/subresources_test.go b/staging/src/k8s.io/apiextensions-apiserver/test/integration/subresources_test.go
index bafcaa28404..deea5f01713 100644
--- a/staging/src/k8s.io/apiextensions-apiserver/test/integration/subresources_test.go
+++ b/staging/src/k8s.io/apiextensions-apiserver/test/integration/subresources_test.go
@@ -23,6 +23,7 @@ import (
 	"sort"
 	"strings"
 	"testing"
+	"time"
 
 	autoscaling "k8s.io/api/autoscaling/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
@@ -691,6 +692,7 @@ func TestGeneration(t *testing.T) {
 			if updatedStatusInstance.GetGeneration() != 1 {
 				t.Fatalf("updating status should not increment .metadata.generation: expected: %v, got: %v", 1, updatedStatusInstance.GetGeneration())
 			}
+			time.Sleep(10 * time.Second)
 
 			gottenNoxuInstance, err = noxuResourceClient.Get("foo", metav1.GetOptions{})
 			if err != nil {
diff --git a/staging/src/k8s.io/client-go/rest/config.go b/staging/src/k8s.io/client-go/rest/config.go
index fb81fb7b1cf..4e53ae87e2a 100644
--- a/staging/src/k8s.io/client-go/rest/config.go
+++ b/staging/src/k8s.io/client-go/rest/config.go
@@ -41,8 +41,8 @@ import (
 )
 
 const (
-	DefaultQPS   float32 = 5.0
-	DefaultBurst int     = 10
+	DefaultQPS   float32 = 5000.0
+	DefaultBurst int     = 10000
 )
 
 var ErrNotInCluster = errors.New("unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined")
diff --git a/staging/src/k8s.io/client-go/tools/cache/reflector.go b/staging/src/k8s.io/client-go/tools/cache/reflector.go
index 8abde7131a3..dd256fe7da9 100644
--- a/staging/src/k8s.io/client-go/tools/cache/reflector.go
+++ b/staging/src/k8s.io/client-go/tools/cache/reflector.go
@@ -330,6 +330,7 @@ loop:
 		case err := <-errc:
 			return err
 		case event, ok := <-w.ResultChan():
+			time.Sleep(50 * time.Millisecond)
 			if !ok {
 				break loop
 			}
```
```
make test-integration WHAT=./staging/src/k8s.io/apiextensions-apiserver/test/integration KUBE_TEST_ARGS="-run ^TestGeneration" GOFLAGS="-v"
--- FAIL: TestGeneration (11.78s)
    testserver.go:141: runtime-config=map[api/all:true]
    testserver.go:142: Starting apiextensions-apiserver on port 46019...
    testserver.go:160: Waiting for /healthz to be ok...
    subresources_test.go:690: unable to update status: Operation cannot be fulfilled on noxus.mygroup.example.com "foo": StorageError: invalid object, Code: 4, Key: /ca6b7607-9396-4317-b5cb-8627b2b95e50/mygroup.example.com/noxus/not-the-default/foo, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: cd619d6f-f115-4ed4-b2fa-75cc8d4b742d, UID in object meta: 63839447-b154-4903-bd56-b51615002d0d
```

as @liggitt suggested preconditions should retry on stale watch data instead of surfacing such error, this is not a test issue. I will send a fix

--- a/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go
@@ -1899,6 +1899,54 @@ func TestDeleteWithCachedObject(t *testing.T) {
 	}
 }
 
+func TestPreconditionalUpdateWithCachedObject(t *testing.T) {
+	podName := "foo"
+	pod := &example.Pod{
+		ObjectMeta: metav1.ObjectMeta{Name: podName},
+		Spec:       example.PodSpec{NodeName: "machine"},
+	}
+	ctx := genericapirequest.WithNamespace(genericapirequest.NewContext(), "test")
+	destroyFunc, registry := newTestGenericStoreRegistry(t, scheme, false)
+	defer destroyFunc()
+
+	// cached object has old UID
+	oldPod, err := registry.Create(ctx, pod, rest.ValidateAllObjectFunc, &metav1.CreateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	registry.Storage.Storage = &staleGuaranteedUpdateStorage{Interface: registry.Storage.Storage, cachedObj: oldPod}
+
+	// delete and re-create the same object with new UID
+	_, _, err = registry.Delete(ctx, podName, rest.ValidateAllObjectFunc, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	obj, err := registry.Create(ctx, pod, rest.ValidateAllObjectFunc, &metav1.CreateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	newPod, ok := obj.(*example.Pod)
+	if !ok {
+		t.Fatalf("unexpected object: %#v", obj)
+	}
+
+	// update the object should not fail precondition
+	newPod.Spec.NodeName = "machine2"
+	res, _, err := registry.Update(ctx, podName, rest.DefaultUpdatedObjectInfo(newPod), rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc, false, &metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// the update should have succeeded
+	r, ok := res.(*example.Pod)
+	if !ok {
+		t.Fatalf("unexpected update result: %#v", res)
+	}
+	if r.Spec.NodeName != "machine2" {
+		t.Fatalf("unexpected, update didn't take effect: %#v", r)
+	}
+}
+
 // TestRetryDeleteValidation checks if the deleteValidation is called again if
 // the GuaranteedUpdate in the Delete handler conflicts with a simultaneous
 // Update.

--- a/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
@@ -274,7 +274,20 @@ func (s *store) GuaranteedUpdate(
 	transformContext := authenticatedDataString(key)
 	for {
 		if err := preconditions.Check(key, origState.obj); err != nil {
-			return err
+			// If our data is already up to date, return the error
+			if !mustCheckData {
+				return err
+			}
+
+			// It's possible we were working with stale data
+			// Actually fetch
+			origState, err = getCurrentState()
+			if err != nil {
+				return err
+			}
+			mustCheckData = false
+			// Retry
+			continue
 		}
 
 		ret, ttl, err := s.updateState(origState, tryUpdate)

